### PR TITLE
Improve annotation API

### PR DESCRIPTION
**Pull request by ryantxu**
This issue aims to improve the annotations API

- [x] Add `created` column to the SQL #11328.  
- [x] Add `modified` column
- [x] Search by userId #11339 
- [x] Search for user created annotations (type=='annotation') #11338 
- [x] Save epoch with ms precision, and migrate existing values #11345
- [x] ~~Decide `sort` syntax.  SQL options are `time ASC`, `time DESC`, `created ASC`,  `created ASC`.  sort is not exposed in other services, so nothing to copy.  Should we punt on `sort` for now?  and tackle that in a new issue?~~ moved to #11344
- [x] Add tests
- [x] Update docs & CHANGELOG

Punt on: Fail when creating an annotation without a dashboard. related to #9939

Note: existing annotations will have a `created=0` and `updated=0`

### CODE REVIEW

**Comment by daniellee** [comment]
My thoughts:

- Adding created and updated columns is totally reasonable. :heavy_check_mark: 
- Search by userId :heavy_check_mark: 
- Search for user created annotations -> would prefer not to use a comparison with zero. Adding a type field would be better. :heavy_check_mark: 
- Can you explain the reasoning behind sorting? Don't really understand the use case for it.

**Comment by ryantxu** [comment]
Great -- I will improve the pull request and ping you when it is worth looking at.

Sorting:  We are using the annotations to have mechanical engineers comment and highlight information in historical data.  I want to see these comments in a list of who created them when.  Think of this as a commenting system where I want to see recent comments.

I will remove sorting from this issue to keep things simple.  For now I can have the client select 1000 values and sort in the client, but it does feel appropriate to include in the API.  I would recommend the same syntax and feature for the dashboards API too 😃

**Comment by ryantxu** [comment]
ok @daniellee, I think it is ready for review.

This is my first attempt at go (not counting a painful attempt at a influx patch), so I'm just stabbing in the dark.  It seems to work though!

**Comment by torkelo** [comment]
Like the  change to ms epoch and the migration, was a mistake on my end to do second epochs

**Comment by bergquist** [comment]
@ryantxu For future ref. don't update the Changelog in the PR. Whoever merges to PR is responsible for that. I'll make sure to update our contributing.md

File: pkg/services/sqlstore/migrations/annotation_mig.go [hunk 1]
```diff
@@ -90,4 +90,29 @@ func addAnnotationMig(mg *Migrator) {
...
+		Cols: []string{"org_id", "created"}, Type: IndexType,
+	}))
+	mg.AddMigration("Add index for updated in annotation table", NewAddIndexMigration(table, &Index{
+		Cols: []string{"org_id", "updated"}, Type: IndexType,
+	}))
+
+	//
+	// Convert epoch saved as seconds to miliseconds
+	//
+	updateEpochSql := "UPDATE annotation SET epoch = (epoch*1000)"
```
Comment (daniellee) [inline pkg/services/sqlstore/migrations/annotation_mig.go]: To be on the safe side, should be something like this:

```sql
UPDATE annotation SET epoch = (epoch*1000) where epoch < 9999999999
```

File: pkg/services/sqlstore/migrations/annotation_mig.go [hunk 2]
```diff
@@ -90,4 +90,29 @@ func addAnnotationMig(mg *Migrator) {
...
+	mg.AddMigration("Convert existing annotations from seconds to miliseconds", new(RawSqlMigration).
```
Comment (daniellee) [inline pkg/services/sqlstore/migrations/annotation_mig.go]: spelling: milliseconds

**Review by daniellee — CHANGES_REQUESTED** [review]

**Comment by ryantxu** [comment]
@daniellee good catch with annotations!  I think i have made the requested changes.  I have not done much testing with the API though, so good to wait if 5.1 is on its way out :)

**Comment by torkelo** [comment]
looks good. 

1) Think for the graphite format cmd.When is in seconds so needs to be multipled by 1000 here https://github.com/grafana/grafana/blob/master/pkg/api/annotations.go#L166

**Review by daniellee — APPROVED** [review]

**Comment by daniellee** [comment]
Looks good to me - I think we will include it in the 5.2 release.

**Pull request merged**